### ISSUE TITLE
unexpected EOF errors when unpacking a large archive

### ISSUE THREAD

**Issue by ncw** [CalebQ42/squashfs#27]
When I use this program to unpack an archive (I wrote this to learn the squashfs API)

```go
package main

import (
	"context"
	"fmt"
	"io"
	"io/fs"
	"log"
	"os"
	"path/filepath"
	"runtime"

	"github.com/CalebQ42/squashfs"
	"golang.org/x/sync/errgroup"
)

func unpackSquashfs(inputFile, outputDir string) (err error) {
	ctx := context.Background()
	f, err := os.Open(inputFile)
	if err != nil {
		return fmt.Errorf("squashfs open failed: %w", err)
	}
	defer func() {
		closeErr := f.Close()
		if err == nil {
			err = closeErr
		}
	}()

	rd, err := squashfs.NewReader(f)
	if err != nil {
		return fmt.Errorf("squashfs reader creation failed: %w", err)
	}

	g, gCtx := errgroup.WithContext(ctx)
	g.SetLimit(runtime.NumCPU())

	err = fs.WalkDir(rd.FS, ".", func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		outPath := filepath.Join(outputDir, path)
		fmt.Printf("Writing %q\n", path)

		if d.IsDir() {
			return os.MkdirAll(outPath, 0777)
		}

		// Run file copies in parallel
		g.Go(func() error {
			if gCtx.Err() != nil {
				return nil
			}

			r, err := rd.FS.Open(path)
			if err != nil {
				return fmt.Errorf("unpack open source failed: %w", err)
			}
			defer r.Close()

			w, err := os.Create(outPath)
			if err != nil {
				return fmt.Errorf("unpack open destination failed: %w", err)
			}
			defer w.Close()

			_, err = io.Copy(w, r)
			if err != nil {
				return fmt.Errorf("unpack copy failed: %w", err)
			}
			return nil
		})

		return nil
	})
	if err != nil {
		return fmt.Errorf("squashfs read walk failed: %w", err)
	}

	err = g.Wait()
	if err != nil {
		return fmt.Errorf("squashfs copies failed: %w", err)
	}

	return nil

}

func main() {
	if len(os.Args) < 3 {
		log.Printf("Syntax: %s <image.sqfs> <directory>", os.Args[0])
		os.Exit(1)
	}
	imageFile, inputDirectory := os.Args[1], os.Args[2]

	err := unpackSquashfs(imageFile, inputDirectory)
	if err != nil {
		log.Fatalf("Failed: %v", err)
	}
}
```

then sometimes I get

    Failed: squashfs read walk failed: readdir usr/local/lib/python3.11/dist-packages/tensorflow/include/external/gemmlowp/meta: unexpected EOF

I can reproduce this with `go-unsquashfs` but it appears to have a memory leak so unless you have a very large amount of RAM it won't finish. This probably deserves another issue!

Exactly which file returns the error varies

    Failed: squashfs read walk failed: readdir usr/local/lib/python3.11/dist-packages/grpc/_cython: unexpected EOF

I tracked this error down to the `binary.Read` returning ErrUnexpectedEOF here

https://github.com/CalebQ42/squashfs/blob/fcd8c4c85bb481989621fee5ea43abc0576b610a/internal/directory/directory.go#L52

It does appear to be reading near the end of the file at the time the unexpected EOF is generated so I conjecture there is some padding at the end or something which is causing the `binary.Read` to read a few extra bytes.

I don't know why it is inconsistent though.

You can generate a squashfs which demonstrates the problem with this script

```bash
#! /usr/bin/bash
ORG=${ORG:-tensorflow}
IMG=${IMG:-tensorflow}
TAG=${TAG:-latest-gpu-jupyter}
docker export $(docker create $ORG/$IMG:$TAG) -o $IMG.tar.gz
mkdir -p $IMG && tar xf $IMG.tar.gz -C $IMG
[ -f $IMG.sqfs ] && rm $IMG.sqfs
mksquashfs $IMG $IMG.sqfs  -comp zstd -Xcompression-level 3 -b 1M -no-xattrs -all-root
```

If that doesn't demonstrate the problem I can send you one which does, but it is 3GB so will take a long time to upload with my puny internet!

PS Actually with the magic of docker you can probably generate an identical archive if I tell you the ID of the image I used is `a2cf87758fef`

**Comment by CalebQ42** [comment]
The RAM issue is known, it's not technically not a leak it's just using all of it, lol. Right now the libraries extract method is not aware of how much ram the system has or how much it's using, causing the out of control RAM usage. I'll eventually add the ability to limit the amount of threads and RAM used , but it'll take a decent amount of work.

As for the actual issue, I have a feeling something deeper is going on if the directory decoding is throwing the error; I'm looking into it now.

**Comment by ncw** [comment]
I managed to upload the problem file eventually [tensorflow.sqfs](https://drive.google.com/open?id=1Ho01W5H_UYibK599qXMKRUOU1y51Mecp) if you need it.

**Comment by ncw** [comment]
BTW I think I discovered where all the memory is going when trying to unpack this archive. This ramps up very quickly but I caught it at only 600MB!

![image](https://github.com/CalebQ42/squashfs/assets/536803/e653ff12-b506-4546-b63e-7e388d3f31a9)

I'm thinking that the zstd decoders aren't having their `Close()` methods called.

**Comment by CalebQ42** [comment]
I finally got around to doing some checking this out (sorry for the wait). The problem seems to be that my library and `fs.WalkDir` don't seem to work together for some reason. The code you provided causes issues even on the archive I use for testing.

For the performance issues with the `Extract` method I tried a couple rudimentary fixes, but nothing seemed to work so it'll take more time to figure out the cause and solution.

**Comment by CalebQ42** [comment]
I started a nearly full re-write of the library to help with #25, and as part of that I completely re-did the individual file's extraction to make it properly limited and more memory friendly and it's not exhibiting the out of control CPU and RAM usage. It'll probably still be a few days before a proper release is out, but wanted to keep you up to date on the progress.

**Comment by CalebQ42** [comment]
Fixed as of `v1.0`. Using `*File.Extract` no longer eats all RAM and has configurable CPU usage. The issues encountered with `fs.WalkDir` seems to also be resolved.
